### PR TITLE
feat: add configuration for file permission

### LINKED ISSUE #712: Set file-system permissions for mounted secrets
**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Ref: https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/722

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

The proto definition has file permission which we set to the default value today:
https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/24b0767489bbd58b487624878a09f3dec916573d/pkg/server/server.go#L73

- If user specifies file permission, we validate and use that
- If no file permissions specified, use the default permission `0644`

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Azure Key Vault provider version: (use the image tag):
- Kubernetes version: (use `kubectl version`):
- Cluster type: (e.g. AKS, aks-engine, etc):

### PR DESCRIPTION
**Pull request by nilekhc**
**Reason for Change**:

This PR adds support for file permission for secrets being written to the pod.

**Requirements**

- [x] squashed commits
- [ ] included documentation
- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

fixes #712 

**Does this change contain code from or inspired by another project?**

- [ ] Yes
- [x] No

**If "Yes," did you notify that project's maintainers and provide attribution?**

**Special Notes for Reviewers**:

### CODE REVIEW

File: pkg/provider/provider.go [hunk 1]
```diff
@@ -98,6 +98,15 @@ type KeyVaultObject struct {
...
 	ObjectEncoding string `json:"objectEncoding" yaml:"objectEncoding"`
+	// Permission is the file permissions
+	Permission string `json:"permission" yaml:"permission"`
+}
+
+// SecretFile holds content and metadata of a secret file
+type SecretFile struct {
+	Content  []byte `json:"content" yaml:"content"`
+	Path     string `json:"path" yaml:"path"`
+	FileMode string `json:"fileMode" yaml:"fileMode"`
```
Comment (aramase) [inline pkg/provider/provider.go:109]: The fields in this struct aren't used for encoding. So we can drop the tags here?

File: pkg/provider/provider.go [hunk 2]
```diff
@@ -98,6 +98,15 @@ type KeyVaultObject struct {
 	// The encoding of the object in KeyVault
 	// Supported encodings are Base64, Hex, Utf-8
 	ObjectEncoding string `json:"objectEncoding" yaml:"objectEncoding"`
+	// Permission is the file permissions
+	Permission string `json:"permission" yaml:"permission"`
```
Comment (aramase) [inline pkg/provider/provider.go:102]: How about `filePermission` for the field name? That's more specific on what the field is intended for.

File: pkg/provider/provider.go [hunk 3]
```diff
@@ -307,7 +316,11 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
 		}
 
 		// these files will be returned to the CSI driver as part of gRPC response
-		files[fileName] = objectContent
+		files = append(files, SecretFile{
+			Path:     fileName,
+			Content:  objectContent,
+			FileMode: keyVaultObject.Permission,
```
Comment (aramase) [inline pkg/provider/provider.go:322]: We can set the default file permission we get from the driver if it's not overridden by the user with `filePermission`

File: pkg/server/server.go [hunk 4]
```diff
@@ -66,12 +67,25 @@ func (s *CSIDriverProviderServer) Mount(ctx context.Context, req *v1alpha1.Mount
...
-			Contents: v,
-			Mode:     int32(filePermission),
-		})
+	for _, file := range files {
+		secretFile := &v1alpha1.File{
+			Path:     file.Path,
+			Contents: file.Content,
+		}
+
+		if file.FileMode != "" {
```
Comment (aramase) [inline pkg/server/server.go:76]: If we set the default file permission from driver when not overridden by user we can get rid of this check

File: pkg/server/server.go [hunk 5]
```diff
@@ -66,12 +67,25 @@ func (s *CSIDriverProviderServer) Mount(ctx context.Context, req *v1alpha1.Mount
...
+			permission, parseErr := strconv.ParseInt(file.FileMode, 8, 32)
```
Comment (aramase) [inline pkg/server/server.go:77]: I think we should validate the file permission before fetching the secrets from keyvault to fail fast. This way we avoid making additional API calls when the final result is going to be an error.

**Review by aramase — COMMENTED** [review]
Added few comments! To summarize,

1. If no file permission is provided, we should default to the permission provided by the driver (same as today). This can be done when the objects to retrieve are unmarshalled.
2. The validation for file permission should happen before actual call to keyvault to avoid making keyvault API calls when final result is an error.

Let's move the validation to a function so we can also unit test it

File: pkg/provider/provider.go (hunk 1 above)
Comment (nilekhc) [reply pkg/provider/provider.go:109]: yup. I was trying different approach for octal conversion which I eventually dropped and forgot to remove this. Now removed.

File: pkg/provider/provider.go (hunk 2 above)
Comment (nilekhc) [reply pkg/provider/provider.go:102]: Sure.. Thought the same but hoping to keep name consistent with driver mount request. I am ok with either name.

**Review by aramase — COMMENTED** [review]
Could we also an e2e test to validate this change?

**Comment by nilekhc** [comment]
> Could we also an e2e test to validate this change?

I have added e2e tests for kind cluster.

**Review by aramase — COMMENTED** [review]
The changes look good! Could we also update the documentation [here](https://azure.github.io/secrets-store-csi-driver-provider-azure/getting-started/usage/#create-your-own-secretproviderclass-object) to add the new field? You can add a `available for version > v1.1.0` so users know it's not yet released.

**Comment by nilekhc** [comment]
> The changes look good! Could you update the documentation [here](https://azure.github.io/secrets-store-csi-driver-provider-azure/getting-started/usage/#create-your-own-secretproviderclass-object) to add the new field? You can add `available for version > v1.1.0` so users know it's not yet released.

added.

File: website/content/en/getting-started/usage/_index.md [hunk 6]
```diff
@@ -51,6 +51,7 @@ To provide identity to access key vault, refer to the following [section](#provi
             objectAlias: SECRET_1           # [OPTIONAL available for version > 0.0.4] object alias
             objectType: secret              # object types: secret, key or cert. For Key Vault certificates, refer to https://azure.github.io/secrets-store-csi-driver-provider-azure/configurations/getting-certs-and-keys/ for the object type to use
             objectVersion: ""               # [OPTIONAL] object versions, default to latest if empty
+            permission: 0755                # [OPTIONAL] permission for secret file being mounted into the pod, default is 0644 if not specified.
```
Comment (aramase) [inline website/content/en/getting-started/usage/_index.md:54]: Isn't this supposed to be [filePermission](https://github.com/Azure/secrets-store-csi-driver-provider-azure/pull/751/files#diff-334716bb990e31490cd8bb9bee9c34757d228c884e35663d88bd26889ffa105bR102)?

File: website/content/en/getting-started/usage/_index.md [hunk 7]
```diff
@@ -76,6 +77,7 @@ To provide identity to access key vault, refer to the following [section](#provi
   | objectVersion          | no       | version of a Key Vault object, if not provided, will use latest                                                                                                                                                        | ""            |
   | objectFormat           | no       | [__*available for version > 0.0.7*__] the format of the Azure Key Vault object, supported types are pem and pfx. `objectFormat: pfx` is only supported with `objectType: secret` and PKCS12 or ECC certificates        | "pem"         |
   | objectEncoding         | no       | [__*available for version > 0.0.8*__] the encoding of the Azure Key Vault secret object, supported types are `utf-8`, `hex` and `base64`. This option is supported only with `objectType: secret`                      | "utf-8"       |
+  | permission         | no       | [__*available for version > v1.1.0*__] permission for secret file being mounted into the pod                      | "0644"       |
```
Comment (aramase) [inline website/content/en/getting-started/usage/_index.md:80]: ditto (`filePermission` instead of `permission`?)

**Review by aramase — APPROVED** [review]

**Pull request merged**